Chapter-4-applying-tactical-domain-driven-design/README.adoc

```diff
@@ -40,29 +77,238 @@ IMPORTANT: It makes no sense to consider the Domain Model in typical CRUD module
-5. Annex can only be attached if _Binding Contract_ is active - has not yet expired or was not terminated.
+2. After the _Contract_ is signed, a __Binding Contract__ is created. It has the force of law and binds the customer with us.
+3. After three months, at any time, the customer can terminate __Binding Contract__ without any penalty.
+4. It is possible to attach an _Annex_ to the existing __Binding Contract__. This way, the customer can extend the contract for another year without preparing a new _Contract_ for him.
+5. Annex can only be attached if a __Binding Contract__ is active - has not yet expired or was not terminated.
 
-image::Assets/flow.jpg[]
+image::Assets/flow.png[]
 
 === Main assumptions
 
 The assumptions remain unchanged to keep the environment comparable to the previous step.
 
 === Solution
 
-==== Overview
+== Overview
 
-In this step we do not change the project structure of the application. We focus only on refactoring the code of _Contracts_ microservice.
+In this step, we do not change the project structure of the application. We focus only on implementing new features and refactoring the code of the _Contracts_ microservice.
 
 We introduce elements like:
 
-- Aggregates
-- Entities
-- Value Objects
-- Domain Events
+- <<aggregate-root,Aggregates>>
+- <<entity,Entities>>
+- <<value-object,Value Objects>>
+- <<domain-events,Domain Events>>
+
+=== Contracts Bounded Context Tactical Design Analysis
+
+image::Assets/design-level-event-storming.png[design level event storming]
+
+Above image is the result of the Event Storming Design Level workshop.
+
+Let's focus on business rules that we identified during the workshop.
+
+1. **Annex Can Only Start During Binding Contract Period**:
+   - An annex can only be attached if it falls within the active period of the ___Binding Contract___. This ensures that all extensions and modifications are valid within the contract's timeframe.
+
+2. **Annex Can Only Be Attached To Active Contracts**:
+   - The ___Binding Contract___ must be active, meaning it hasn't expired or been terminated. This rule prevents any modifications to contracts that are no longer valid.
+
+3. **Previous Annex Must Be Signed**:
+   - Any new annex can only be added if the previous annex has been signed. This maintains a clear and enforceable order of amendments, ensuring that no annex is added without proper authorisation.
+
+Let's take a look closer to this rule. __Previous Annex Must Be Signed__. This rule is connected with the relationship between the new and previous annex. This is invariant.
+
+NOTE: Invariant is a rule or condition that must always be true for a system to be considered in a valid state. It ensures the integrity and consistency of the domain model
+
+To enforce these business rules and maintain consistency, we need a robust way to protect invariant, especially in a concurrent environment.
+
+This is where the concept of an <<aggregate-root,Aggregate Root>> comes into play.
```

Comment:
Aggregate, not aggregate root. The root of the aggregate is the entity that is promoted. Itself, the aggregate is responsible of guarding the consistency

Reply:
Alright, it's good to know this distinction, thank you.